drivers/vdpa/vfe/vfe_vdpa.c

```diff
@@ -0,0 +1,1196 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ * Copyright 2019 Mellanox Technologies, Ltd
+ */
+#include <unistd.h>
+#include <net/if.h>
+#include <sys/socket.h>
+#include <sys/ioctl.h>
+#include <fcntl.h>
+#include <netinet/in.h>
+
+#include <rte_malloc.h>
+#include <rte_log.h>
+#include <rte_errno.h>
+#include <rte_string_fns.h>
+#include <rte_bus_pci.h>
+#include <rte_vfio.h>
+#include "rte_vhost.h"
+#include <rte_vdpa.h>
+#include <vdpa_driver.h>
+
+#include <rte_kvargs.h>
+
+#include "virtio_pci.h"
+#include "virtqueue.h"
+
+RTE_LOG_REGISTER(vfe_vdpa_logtype, pmd.vdpa.vfe, NOTICE);
+#define DRV_LOG(level, fmt, args...) \
+	rte_log(RTE_LOG_ ## level, vfe_vdpa_logtype, \
+		"VFE %s(): " fmt "\n", __func__, ##args)
+
+struct virtio_hw_internal vfe_virtio_hw_internal[VFE_MAX_PORT_NUM];
+
+enum {
+	VFE_VDPA_NOTIFIER_STATE_DISABLED,
+	VFE_VDPA_NOTIFIER_STATE_ENABLED,
+	VFE_VDPA_NOTIFIER_STATE_ERR
+};
+
+struct vfe_vring_info {
+	uint64_t desc;
+	uint64_t avail;
+	uint64_t used;
+	uint16_t size;
+	uint16_t last_avail_idx;
+	uint16_t last_used_idx; /*other is for future use, not use now*/
+	bool enable;
+	uint8_t notifier_state;
+	uint16_t index;
+	struct rte_intr_handle *intr_handle;
+	struct vfe_vdpa_priv *priv;
+};
+
+#define VFE_VDPA_DRIVER_NAME vdpa_vfe
+struct vfe_vdpa_priv {
+	TAILQ_ENTRY(vfe_vdpa_priv) next;
+	struct rte_pci_device *pdev;
+	struct rte_vdpa_device *vdev;
+	struct virtio_pci_dev vpdev;
+	int vfio_container_fd;
```

Comment:
all the 3 fields does belong here, and also pdev on line 56.

This should be in virtio_pci_dev structure.
have virtio_pci_dev structure pointer here.